aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs

```diff
@@ -4,58 +4,118 @@
-// we'd have to integrate more closely with the orchestrator.
-//
-// I'll leave this test here because we do eventually want to support stalled
-// stream protection for uploads.
-#[ignore]
+enum SlowBodyState {
+    Wait(Pin<Box<dyn std::future::Future<Output = ()> + Send + Sync + 'static>>),
+    Send,
+    Taken,
+}
+
+struct SlowBody {
+    state: SlowBodyState,
+}
+
+impl SlowBody {
+    fn new() -> Self {
+        Self {
+            state: SlowBodyState::Send,
+        }
+    }
+}
+
+impl Body for SlowBody {
+    type Data = Bytes;
+    type Error = BoxError;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        loop {
+            let mut state = SlowBodyState::Taken;
+            std::mem::swap(&mut state, &mut self.state);
+            match state {
+                SlowBodyState::Wait(mut fut) => match fut.as_mut().poll(cx) {
+                    Poll::Ready(_) => self.state = SlowBodyState::Send,
+                    Poll::Pending => {
+                        self.state = SlowBodyState::Wait(fut);
+                        return Poll::Pending;
+                    }
+                },
+                SlowBodyState::Send => {
+                    self.state = SlowBodyState::Wait(Box::pin(sleep(Duration::from_micros(100))));
+                    return Poll::Ready(Some(Ok(Bytes::from_static(
+                        b"data_data_data_data_data_data_data_data_data_data_data_data_\
+                          data_data_data_data_data_data_data_data_data_data_data_data_\
+                          data_data_data_data_data_data_data_data_data_data_data_data_\
+                          data_data_data_data_data_data_data_data_data_data_data_data_",
+                    ))));
+                }
+                SlowBodyState::Taken => unreachable!(),
+            }
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        Poll::Ready(Ok(None))
+    }
+}
+
 #[tokio::test]
 async fn test_stalled_stream_protection_defaults_for_upload() {
-    // We spawn a faulty server that will close the connection after
-    // writing half of the response body.
+    let _logs = capture_test_logs();
+
+    // We spawn a faulty server that will stop all request processing after reading half of the request body.
     let (server, server_addr) = start_faulty_upload_server().await;
     let _ = tokio::spawn(server);
 
     let conf = Config::builder()
         .credentials_provider(Credentials::for_tests())
         .region(Region::new("us-east-1"))
         .endpoint_url(format!("http://{server_addr}"))
-        // .stalled_stream_protection(StalledStreamProtectionConfig::enabled().build())
+        .stalled_stream_protection(StalledStreamProtectionConfig::enabled().build())
```

Comment:
by default is it not enabled for upload? why is this required?

Reply:
Just being explicit.